### ISSUE TITLE
Add `using` option to explicitly specify the strategy to use

### ISSUE THREAD

**Issue by samford** [Homebrew/homebrew-livecheck#301]
**NOTE**: There are other feature proposals to come and this primarily applies to niche scenarios, so I wouldn't recommend that anyone take this on just yet. There are more useful features that should be addressed first. I've only created an issue here now because I wanted to avoid bogging down the livecheck DSL issue in Homebrew/brew with too much detail on `using`.

---

The basic idea here is to restructure livecheck's heuristic so it can accept a `:using` option to explicitly specify the strategy to use for a given formula.  The value would be a symbol, corresponding to a strategy in the heuristic (e.g., `:git`, `:sourceforge`, etc.), or an array of these symbols (which livecheck would choose from instead of the pool of all possible strategies).

Though we could have values like `:git` or `:sourceforge`, I imagine this will be most useful for specifying that the default strategy (`page_matches`) should be used instead of a specific strategy. That default strategy would need to be named somehow like `:default` or `:page`.  Some people may erroneously think that `:default` means "the strategy that livecheck would normally choose for this URL", so something like `:page` may be better.

`using` should work both with or without a URL.  When a URL is provided, livecheck would use the specified strategy without going through the built-in strategy-guessing (based on the URL).  When a URL isn't provided, livecheck would gather the URLs from the formula (like it already does) but only use the specified strategy (or strategies) when checking each URL.

It looks like this in practice:

```ruby
class SomeSoftware
  livecheck :url   => "https://www.example.com/downloads",
            :using => :page,
            :regex => /somesoftware-(\d+(?:\.\d+)+)\.tar/
end
```

Assuming we had a specific strategy for example.com URLs, this would override it and use `page_matches` instead with the regex.

The two main applications I imagine for `using` are detailed below.

## Overriding `livecheck`'s Built-In Strategy Selection

`livecheck` does two types of guessing when looking for versions:

* If a URL isn't provided in a livecheckable, it goes through the URLs in the formula (in order: HEAD, stable, stable mirrors, homepage) until it gets a result.
* With each URL, it decides what strategy to apply to the URL by going through a big `if` statement.

We can already tell livecheck to only check a specific URL but it decides for itself what strategy to use.

At least once before, I encountered a situation where there was an exception to the condition where a strategy should be applied. Based on the URL provided, livecheck was using a specific strategy and there was no way to use the default `page_matches` strategy (as it comes after the specific strategy in the `if-elsif` chain). Luckily, I was able to modify the condition for the aforementioned strategy so it wouldn't apply (while still working properly for others). However, if I wasn't able to do that, it wouldn't have been possible to create a working livecheckable.

In this example, `using` would force livecheck to only use the provided strategy and not do any strategy guessing. The livecheckable here would have contained `url` (restricting livecheck to one URL), `using` (restricting livecheck to one strategy), and `regex` (determining what text to match on the page).

## Restricting the Formula URLs `livecheck` Uses, by Strategy

Another application of `using` could be filtering out URLs that a strategy doesn't apply to, so livecheck does less work (or loudly fails rather than quietly running the regex on an inappropriate target). This isn't likely to be used very often but it requires almost no extra work to implement beyond what's already required for this feature.

Implementing `using` will likely require us to pull out the code for each strategy into methods, so we can run those methods when a strategy is supplied via `using`. Similarly, we could refactor the default strategy-selection code (the big `if` statement) into a method that takes the URL and returns the strategy to use. When `using` isn't provided, the strategy-selection method would be run on a given URL and afterward the associated strategy would be executed.

This not only helps to clean up the heuristic (making it easier to read and understand) but also allows us to implement the aforementioned behavior when `using` is provided in a livecheckable but a `url` is not. Basically, running the strategy-selection code tells us what strategy would be used with the URL by default and we can just add something to skip the URL like `next if using && strategy != using` before the code that executes the strategy.

## Closing

As mentioned above, this is relatively low priority and should be handled after other forthcoming feature proposals.  This is primarily meant to be a catch-all for any specific discussion of the `using` feature that doesn't belong in the livecheck DSL issue.

**Comment by samford** [comment]
I'm hoping to work on this sometime in the coming weeks, so please don't close this stale bot.

**Comment by samford** [comment]
Just as an update, #1025 finishes laying the groundwork for the `strategy` feature (previously referred to as `using` here) and I will be creating a PR for the feature immediately after #1025 is merged. The work has been done for over a month now but I'm just now getting around to creating PRs to wrap this up.
